File: drills/ts/solutions/nc122_swim_in_water__binary_search.ts
export function swimInWater(grid: number[][]): number {
  if (grid.length === 0) return 0;

  const n = grid.length;

  const reaches = (limit: number): boolean => {
    // The target has to be passable itself, so its depth is checked before it
    // counts as reached.
    if (grid[0][0] > limit) return false;
    const seen = new Set<string>();
    const stack: [number, number][] = [[0, 0]];
    while (stack.length) {
      const [r, c] = stack.pop()!;
      if (seen.has(`${r},${c}`) || grid[r][c] > limit) continue;
      if (r === n - 1 && c === n - 1) return true;
      seen.add(`${r},${c}`);
      for (const [nr, nc] of [[r - 1, c], [r + 1, c], [r, c - 1], [r, c + 1]]) {
        if (nr >= 0 && nr < n && nc >= 0 && nc < grid[nr].length) stack.push([nr, nc]);
      }
    }
    return false;
  };

  // Reachability at time t is monotone: once the corner can be reached it stays
  // reachable as the water rises further. That is exactly the shape binary
  // search needs, so the question turns from "what is the cheapest path" into
  // "is it possible yet", answered by a plain flood fill.
  let low = grid[0][0];
  let high = n * n - 1;
  while (low < high) {
    const middle = Math.floor((low + high) / 2);
    if (reaches(middle)) high = middle;
    else low = middle + 1;
  }
  return low;
}

File: drills/python/solutions/nc122_swim_in_water__binary_search.py
def swimInWater(grid):
    if not grid:
        return 0

    n = len(grid)
    target = (n - 1, n - 1)

    def reaches(limit):
        # The target has to be passable itself, so its depth is checked before
        # it counts as reached.
        if grid[0][0] > limit:
            return False
        seen = set()
        stack = [(0, 0)]
        while stack:
            r, c = stack.pop()
            if (r, c) in seen or grid[r][c] > limit:
                continue
            if (r, c) == target:
                return True
            seen.add((r, c))
            for nr, nc in ((r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)):
                if 0 <= nr < n and 0 <= nc < len(grid[nr]):
                    stack.append((nr, nc))
        return False

    # Reachability at time t is monotone: once the corner can be reached it
    # stays reachable as the water rises further. That is exactly the shape
    # binary search needs, so the question turns from "what is the cheapest
    # path" into "is it possible yet", answered by a plain flood fill.
    low, high = grid[0][0], n * n - 1
    while low < high:
        middle = (low + high) // 2
        if reaches(middle):
            high = middle
        else:
            low = middle + 1
    return low
